src/tests/tcurl_test_tool.c

```diff
@@ -0,0 +1,222 @@
+        case 'v':
+            pc_verbose = 1;
+            break;
+        default:
+            DEBUG(SSSDBG_FATAL_FAILURE, "Unexpected option\n");
+            return 1;
+        }
+    }
+
+    DEBUG_CLI_INIT(pc_debug);
+
+    tool_ctx = talloc_zero(NULL, struct tool_ctx);
+    if (tool_ctx == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Could not init tool context\n");
+        return 1;
+    }
+
+    inbufs = talloc_zero_array(tool_ctx, struct sss_iobuf *, MAXREQ);
+    if (inbufs == NULL) {
+        talloc_zfree(tool_ctx);
+        return 1;
+    }
+
+    while ((extra_arg_ptr = poptGetArg(pc)) != NULL) {
+        switch(req_type) {
+        case TCURL_HTTP_GET:
+        case TCURL_HTTP_DELETE:
+            urls[n_reqs++] = extra_arg_ptr;
+            break;
+        case TCURL_HTTP_PUT:
+            if (urls[n_reqs] == NULL) {
+                urls[n_reqs] = extra_arg_ptr;
+            } else {
+                inbufs[n_reqs] = sss_iobuf_init_readonly(
+                                              inbufs,
+                                              (uint8_t *) discard_const(extra_arg_ptr),
+                                              strlen(extra_arg_ptr));
+                if (inbufs[n_reqs] == NULL) {
+                    DEBUG(SSSDBG_CRIT_FAILURE, "Could not init input buffer\n");
+                    talloc_zfree(tool_ctx);
+                    return 1;
+                }
+                n_reqs++;
+            }
+            break;
+        }
+    }
+
+    if (opt != -1) {
+        poptPrintUsage(pc, stderr, 0);
+        fprintf(stderr, "%s", poptStrerror(opt));
+        talloc_zfree(tool_ctx);
+        return 1;
+    }
+
+    if (!socket_path) {
+        DEBUG(SSSDBG_FATAL_FAILURE, "Please specify the socket path\n");
+        poptPrintUsage(pc, stderr, 0);
+        talloc_zfree(tool_ctx);
+        return 1;
+    }
+
+    tool_ctx->nreqs = n_reqs;
+    tool_ctx->verbose = !!pc_verbose;
+
+    ev = tevent_context_init(tool_ctx);
+    if (ev == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Could not init tevent context\n");
+        talloc_zfree(tool_ctx);
+        return 1;
+    }
+
+    ctx = tcurl_init(tool_ctx, ev);
+    if (ctx == NULL) {
+        DEBUG(SSSDBG_FATAL_FAILURE, "Could not init tcurl context\n");
+        talloc_zfree(tool_ctx);
+        return 1;
+    }
+
+    for (size_t i = 0; i < n_reqs; i++) {
+        req = tcurl_http_send(tool_ctx, ev, ctx,
+                              req_type,
+                              socket_path,
+                              urls[i],
+                              headers,
+                              inbufs[i],
+                              5);
+        if (ctx == NULL) {
+            DEBUG(SSSDBG_FATAL_FAILURE, "Could not create request\n");
+            talloc_zfree(tool_ctx);
+            return 1;
+        }
+        tevent_req_set_callback(req, request_done, tool_ctx);
+    }
+
+    while (tool_ctx->done == false) {
+         tevent_loop_once(ev);
+    }
+
```

Comment:
I think we should implement some timeout here to prevent deadlocks. (we currently hit on el7)
IMHO it can be hardcoded (1min) or taken from command line argument.